File: ProtocJsGenPlugin/StringExtensions.cs
using System.Text;

namespace ProtocJsGenPlugin;

internal static class StringExtensions
{
    /// <summary>
    /// 将snake_case转换为camelCase
    /// </summary>
    public static string SnakeToCamelCase(this string snakeCase)
    {
        if (string.IsNullOrEmpty(snakeCase))
            return snakeCase;

        var parts = snakeCase.Split('_');
        var result = new StringBuilder();

        for (int i = 0; i < parts.Length; i++)
        {
            var part = parts[i];
            if (string.IsNullOrEmpty(part))
                continue;

            if (i == 0)
            {
                // 第一个单词保持小写
                result.Append(part);
            }
            else
            {
                // 后续单词首字母大写
                result.Append(char.ToUpperInvariant(part[0]));
                if (part.Length > 1)
                    result.Append(part.Substring(1));
            }
        }

        return result.ToString();
    }

    /// <summary>
    /// 将snake_case转换为PascalCase
    /// </summary>
    public static string SnakeToPascalCase(this string snakeCase)
    {
        if (string.IsNullOrEmpty(snakeCase))
            return snakeCase;

        var parts = snakeCase.Split('_');
        var result = new StringBuilder();

        foreach (var part in parts)
        {
            if (string.IsNullOrEmpty(part))
                continue;

            result.Append(char.ToUpperInvariant(part[0]));
            if (part.Length > 1)
                result.Append(part.Substring(1));
        }

        return result.ToString();
    }
}

File: src/string_extensions.cc
#include "string_extensions.h"

#include <cctype>
#include <sstream>
#include <string>
#include <vector>

namespace protoc_js_gen_plugin {
namespace {

std::vector<std::string> SplitString(const std::string& str, char delimiter) {
    std::vector<std::string> parts;
    std::stringstream ss(str);
    std::string part;
    while (std::getline(ss, part, delimiter)) {
        parts.push_back(part);
    }
    return parts;
}

}  // namespace

std::string SnakeToCamelCase(const std::string& snake_case) {
    if (snake_case.empty()) return snake_case;

    auto parts = SplitString(snake_case, '_');
    std::string result;

    for (size_t i = 0; i < parts.size(); ++i) {
        const auto& part = parts[i];
        if (part.empty()) continue;

        if (i == 0) {
            // First word stays lowercase
            result += part;
        } else {
            // Subsequent words start with uppercase
            result += static_cast<char>(std::toupper(static_cast<unsigned char>(part[0])));
            if (part.size() > 1) {
                result += part.substr(1);
            }
        }
    }

    return result;
}

std::string SnakeToPascalCase(const std::string& snake_case) {
    if (snake_case.empty()) return snake_case;

    auto parts = SplitString(snake_case, '_');
    std::string result;

    for (const auto& part : parts) {
        if (part.empty()) continue;

        result += static_cast<char>(std::toupper(static_cast<unsigned char>(part[0])));
        if (part.size() > 1) {
            result += part.substr(1);
        }
    }

    return result;
}

}  // namespace protoc_js_gen_plugin